packages/next/client/script.tsx

```diff
@@ -0,0 +1,155 @@
+import React, { useEffect, useContext } from 'react'
+import { ScriptHTMLAttributes } from 'react'
+import { HeadManagerContext } from '../next-server/lib/head-manager-context'
+
+const ScriptCache = new Set()
+
+const DOMAttributeNames: Record<string, string> = {
+  acceptCharset: 'accept-charset',
+  className: 'class',
+  htmlFor: 'for',
+  httpEquiv: 'http-equiv',
+}
+
+interface Props extends ScriptHTMLAttributes<HTMLScriptElement> {
+  strategy?: 'defer' | 'lazy' | 'dangerousRenderBlocking' | 'eager'
+  key?: string
+  onLoad?: () => void
+  onError?: () => void
+  children?: any
+  preload?: boolean
+}
+
+const onErrorHandler: Function = (
+  cacheKey: string,
+  onError: Function
+) => () => {
+  if (cacheKey) {
+    ScriptCache.delete(cacheKey)
+  }
+
+  if (onError) {
+    onError()
+  }
+}
+
+const loadScript = (props: Props) => {
+  const {
+    src = '',
+    onLoad = () => {},
+    dangerouslySetInnerHTML,
+    children = '',
+    key,
+    onError,
+  } = props
+
+  if (ScriptCache.has(key || src)) {
+    return
+  }
+
+  const el = document.createElement('script')
+  const cacheKey = key || src
+
+  if (cacheKey) {
+    ScriptCache.add(cacheKey)
+  }
+
+  if (dangerouslySetInnerHTML) {
+    el.innerHTML = dangerouslySetInnerHTML.__html || ''
+  } else if (children) {
+    el.textContent = typeof children === 'string' ? children : children.join('')
+  } else if (src) {
+    el.src = src
+  }
+
+  el.onload = onLoad
```

Comment:
I'm imagining a typical use-case for this would be to load an API of an embedded widget (is that a correct assumption?). Something along the lines of:

```js
function MyExternalWidget () {
  const widgetRef = React.useRef()
  const onLoad = React.useCallback(() => {
    window.myEmbedderApi.renderWidget(widgetRef.current)
  })
  return (
    <div>
      <ScriptLoader src='https://my-embedder-api' onLoad={onLoad} />
      <div ref={widgetRef} />
    </div>
  )
}
```

However, this type of code only works if the `onLoad` gets called once for each instance of `ScriptLoader`. If I read this PR correctly, this PR calls `onLoad` only once for all `ScriptLoader` instances with the same `src` during the whole lifecycle of the browser window.

Reply:
Yes. The assumption is that we want to load a third party script only once, and do not trigger a network request on rerenders. 
onLoad as the name suggests should be used to execute some logic after the script loads..

Even in the case of an embedded widget, I presume we only want to call the external script and setup the widget only once? 
In cases where we don't want the src to be counted as the same, we can pass a key prop.

Reply:
> Even in the case of an embedded widget, I presume we only want to call the external script and setup the widget only once?

yes, but if I read this PR correctly, using the example I showed above as:
```jsx
<MyExternalWidget />
<MyExternalWidget />
<MyExternalWidget />
```
will only initialize 1 of these 3 widgets. I agree that the script should be loaded only once, but the `onLoad` event should always fire regardless of whether a script was cached or not.

As far as I know, this is how the original `<script>` tag works, `onLoad` will always fire, regardless of whether the script was fetched or comes from cache

Reply:
I think I understand what you meant. As a start I've made sure the onLoad function is called each time now.

Reply:
> As a start I've made sure the onLoad function is called each time now.

Looks like it would potentially get called before the script is actually loaded. I think you'll need to cache a `Promise` instead, and `await` it before calling `onLoad`. Definitely needs an integration test.